enhancements/network/user-defined-network-segmentation.md

```diff
@@ -0,0 +1,947 @@
+multiple networks, we need to consider potential attack vectors like a tenant trying to exhaust OVN-Kubernetes
+resources by creating too many secondary networks.
+
+Furthermore, in the previous proposal it mentions using an annotation on the namespace to indicate if the network was
+configured by OVN-Kubernetes and is in use for the namespace. By utilizing a CRD, the status of the network CRD itself
+can be used to indicate whether it is configured by OVN-Kubernetes, eliminating the need for annotating a namespace. 
+
+### IP Addressing
+
+As previously mentioned, one of the goals is to allow user-defined networks to have overlapping pod IP addresses. This
+is enabled by allowing a user to configure what CIDR to use for pod addressing when they create the network. However,
+this range cannot overlap with the default cluster CIDR used by the cluster default network today.
+
+Furthermore, the internal masquerade subnet and the Kubernetes service subnet will remain unique and will exist globally
+to serve all networks. The masquerade subnet must be large enough to accommodate enough networks. Therefore, the
+subnet size of the masquerade subnet is equal to the number of desired networks * 2, as we need 2 masquerade IPs per
+network. The masquerade subnet remains localized to each node, so each node can use the same IP addresses and the size
+of the subnet does not scale with number of nodes.
+
+The transit switch subnets may overlap between all networks. This network is just used for transport between nodes, and
+is never seen by the pods or external clients.
+
+The join subnet of the default cluster network may not overlap with the join subnet of user-defined networks. This is
+due to the fact that the pod is connected to the default network, as well as the user-defined primary network. The join
+subnet is SNAT'ed by the GR of that network in order to facilitate ingress reply service traffic going back to the
+proper GR, in case it traverses the overlay. For this reason, the pods may see this IP address and routes are added to
+the pod to steer the traffic to the right interface (100.64.0.0/16 is the default cluster network join subnet):
+
+```
+[root@pod3 /]# ip route show
+default via 10.244.1.1 dev eth0 
+10.96.0.0/16 via 10.244.1.1 dev eth0 
+10.244.0.0/16 via 10.244.1.1 dev eth0 
+10.244.1.0/24 dev eth0 proto kernel scope link src 10.244.1.8 
+100.64.0.0/16 via 10.244.1.1 dev eth0
+```
+
+Since the pod needs routes for each join subnet, any layer 3 network that is attached to the pod needs a unique join
+subnet. A layer 2 network does not use the join subnet, and the GR IP will be used to SNAT ingress service traffic there.
+A pod connected to the default cluster network (100.64.0.0/16), a user-defined, layer 3, primary network (100.65.0.0/16),
+and a layer 2, secondary network (100.66.0.0/16), the routing table would look like:
```

Comment:
Right, you cannot have more than one primaryNetwork NAD connected to a pod, no matter if its layer 2 or layer 3.

Oh, I forgot you were using the join subnet as a way to assign secondary IPs in a different subnet for the GRs. Why did you take that approach?

https://raw.githubusercontent.com/openshift/enhancements/b0175b82a3259ec7f788e9beea6d168e69f1e3c4/enhancements/network/images/multi-homing-l2-gw.svg

Is it to conserve pod network IP addresses? Is that join subnet hidden from the pods (no routes on the pod for join subnet), and just used for re-routing between GRs?

Reply:
At L2 since we don't have a transit switches between GR's we should add a remote port to each GR at the logical switch and that means that we have to put an IP on those ports (join subnet) to be able to route over them if pod traffic need to go to different GR (for example for the egress IP feature).

Reply:
I'm updating the document to indicate that layer 2 networks will also use the join subnet and it must be unique. This creates more consistency that in L3 and L2 they both are unique subnets and are essentially used for the same purpose.